File: Java/LinkedList/01_Singly_Linear_LinkedList_Java/SinglyLL_011.java
/*
================================================================================
Program Name    : Singly Linked List - InsertLast() Implementation
Program Number  : 011 (Java Version)
Description     : This program demonstrates the implementation of
                  InsertLast() in a Singly Linear Linked List. The
                  operation traverses the linked list using a
                  temporary reference until the last node is found,
                  then links the newly created node at the end of
                  the list. The node counter (iCount) is updated
                  automatically after every successful insertion.
Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
================================================================================
Concepts Used:
1. Class Encapsulation (private/public)
2. InsertFirst() Implementation
3. InsertLast() Implementation
4. Linked List Traversal
5. Temporary Reference (temp)
6. Automatic iCount Increment
7. Constructor Pattern
8. Dynamic Object Creation

Key Learning Points:
- InsertLast() inserts a new node at the end of the linked list.
- A temporary reference is used for traversal.
- Traversal continues until temp.next becomes null.
- The last node is linked to the newly created node.
- iCount is updated automatically after every insertion.
- Count() returns the total number of nodes in O(1) time.
- InsertLast() requires traversal, so its time complexity is O(n).
*/
class node
{
    public int data;
    public node next;

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL()
    {
        this.first = null;
        this.iCount = 0;
    }

    public void Display()
    {}

    public int Count()
    {
        return iCount;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);

        if(first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }
        iCount++;
    }

    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);

        if(first == null)
        {
            first = newn;
        }
        else
        {
            node temp = null;
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }

        iCount++;
    }

    public void InsertAtPos(int iNo,int iPos)
    {}

    public void DeleteFirst()
    {}

    public void DeleteLast()
    {}

    public void DeleteAtPos(int iPos)
    {}
}

class SinglyLL011
{
    public static void main(String A[])
    {
        SinglyLL sobj = new SinglyLL();

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);
    }
}
/*
================================================================================
Expected Output:
================================================================================

(No output - builds linked list in memory)

Linked List Created:

first → [11] → [21] → [51] → [101] → [111] → [121] → null

iCount = 6

================================================================================
Memory Diagram:
================================================================================

After InsertFirst() Operations

first
  |
  v
[11|→] → [21|→] → [51|null]

iCount = 3


After InsertLast(101)

first
  |
  v
[11|→] → [21|→] → [51|→] → [101|null]

iCount = 4


After InsertLast(111)

first
  |
  v
[11|→] → [21|→] → [51|→] → [101|→] → [111|null]

iCount = 5


After InsertLast(121)

first
  |
  v
[11|→] → [21|→] → [51|→] → [101|→] → [111|→] → [121|null]

iCount = 6

================================================================================
Pointer/Logic Explanation:
================================================================================

InsertLast() Logic:

Step 1:
Create a new node.

node newn = new node(iNo);

Step 2:
Check whether the linked list is empty.

if(first == null)

If the list is empty,
the new node becomes the first node.

Step 3:
If the list is not empty,
create a temporary reference.

node temp = first;

Step 4:
Traverse the linked list.

while(temp.next != null)
{
    temp = temp.next;
}

The loop stops when temp reaches the last node.

Step 5:
Attach the new node.

temp.next = newn;

Step 6:
Increment the node counter.

iCount++;

Execution Trace:

InsertLast(101)

Current List

11 → 21 → 51 → null

temp starts from first

11 → 21 → 51

temp reaches 51

51.next = 101

Result

11 → 21 → 51 → 101

------------------------------------------------------------

InsertLast(111)

Traversal

11 → 21 → 51 → 101

Last node = 101

101.next = 111

Result

11 → 21 → 51 → 101 → 111

------------------------------------------------------------

InsertLast(121)

Traversal

11 → 21 → 51 → 101 → 111

Last node = 111

111.next = 121

Final List

11 → 21 → 51 → 101 → 111 → 121

================================================================================
Time Complexity:
================================================================================

InsertFirst()    : O(1)

InsertLast()     : O(n)

Count()          : O(1)

Space Complexity : O(1)

================================================================================
Conclusion:
================================================================================

Program 011 demonstrates the implementation of the
InsertLast() operation in a Singly Linear Linked List.
A temporary reference is used to traverse the linked
list until the last node is reached, where the newly
created node is linked successfully. The node counter
is maintained automatically after every insertion,
allowing Count() to execute in constant time O(1).
This program completes the implementation of insertion
operations at both the beginning and the end of the
linked list.

================================================================================
*/

File: C++/LinkedList/01_Singly_Linear_LinkedList_CPP/SinglyLL_011.cpp
/*
===============================================================================

Program Name    : Singly Linked List - InsertFirst() and InsertLast()
Program Number  : 011
Description     : This program demonstrates the implementation of
                  InsertFirst() and InsertLast() operations in a
                  Singly Linked List using C++ classes.

                  InsertFirst() inserts a new node at the beginning
                  of the linked list, while InsertLast() inserts a
                  new node at the end. The program also displays the
                  linked list and maintains the total number of nodes
                  using the Count() function.

Concepts Used   : C++ Classes,
                  Objects,
                  Constructor,
                  Dynamic Memory Allocation (new),
                  Self-Referential Structure,
                  Pointer Manipulation,
                  Encapsulation,
                  InsertFirst(),
                  InsertLast(),
                  Display(),
                  Count()

Input           : Node values inserted through function calls.

Output          : Displays the linked list after InsertFirst() and
                  InsertLast() operations along with the total
                  number of nodes.

Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve

===============================================================================
*/
#include<iostream>
using namespace std;

#pragma pack(1)
struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

class SinglyLL
{
    private:
        PNODE first;
        int iCount;

    public:
        SinglyLL();
        void Display();
        int Count();
        void InsertFirst(int iNo);
        void InsertLast(int iNo);
        void InsertAtPos(int iNo, int iPos);
        void DeleteFirst();
        void DeleteLast();
        void DeleteAtPos(int iPos);
};

SinglyLL::SinglyLL()
{
    this -> first = NULL;
    this -> iCount = 0;
}

void SinglyLL:: Display()
{
    PNODE temp = NULL;
    temp = this -> first;

    while(temp != NULL)
    {
        cout<<"| "<<temp -> data<<" |-> ";
        temp = temp -> next;
    }
    cout<<"NULL"<<endl;
}

int SinglyLL:: Count()
{
    return this -> iCount;
}

void  SinglyLL:: InsertFirst(int iNo)
{
    PNODE newn = NULL;

    newn = new NODE;

    newn -> data = iNo;
    newn -> next = NULL;

    if(this -> first == NULL)
    {
        this -> first = newn;
    }
    else
    {
        newn -> next = this -> first;
        this -> first = newn;
    }

    this -> iCount++;
}

void  SinglyLL:: InsertLast(int iNo)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    newn = new NODE;

    newn -> data = iNo;
    newn -> next = NULL;

    if(this -> first == NULL)
    {
        this -> first = newn;
    }
    else
    {
        temp = this -> first;

        while(temp -> next != NULL)
        {
            temp = temp -> next;
        }

        temp -> next = newn;
    }

    this -> iCount++;
}

void  SinglyLL:: InsertAtPos(int iNo, int iPos)
{}

void  SinglyLL:: DeleteFirst()
{}

void  SinglyLL:: DeleteLast()
{}

void  SinglyLL:: DeleteAtPos(int iPos)
{}

int main()
{
    SinglyLL sobj;
    int iRet = 0;

    sobj.InsertFirst(51);
    sobj.InsertFirst(21);
    sobj.InsertFirst(11);

    sobj.Display();

    iRet = sobj.Count();
    cout<<"Number of Elements are: "<<iRet<<endl;

    sobj.InsertLast(101);
    sobj.InsertLast(111);
    sobj.InsertLast(121);

    sobj.Display();

    iRet = sobj.Count();
    cout<<"Number of Elements are: "<<iRet<<endl;

    return 0;
}
/*
===============================================================================

Expected Output:
===============================================================================

| 11 |-> | 21 |-> | 51 |-> NULL
Number of Elements are: 3

| 11 |-> | 21 |-> | 51 |-> | 101 |-> | 111 |-> | 121 |-> NULL
Number of Elements are: 6

===============================================================================
Memory Representation:
===============================================================================

After InsertFirst(51)

first
 │
 ▼
+------+------+
|  51  | NULL |
+------+------+

-------------------------------------------------------------------------------

After InsertFirst(21)

first
 │
 ▼
+------+-----+     +------+------+
|  21  |  • ─────► |  51  | NULL |
+------+-----+     +------+------+

-------------------------------------------------------------------------------

After InsertFirst(11)

first
 │
 ▼
+------+-----+     +------+-----+     +------+------+
|  11  |  • ─────► |  21  |  • ─────► |  51  | NULL |
+------+-----+     +------+-----+     +------+------+

iCount = 3

-------------------------------------------------------------------------------

After InsertLast(101)

first
 │
 ▼
11 → 21 → 51 → 101 → NULL

-------------------------------------------------------------------------------

After InsertLast(111)

first
 │
 ▼
11 → 21 → 51 → 101 → 111 → NULL

-------------------------------------------------------------------------------

After InsertLast(121)

first
 │
 ▼
11 → 21 → 51 → 101 → 111 → 121 → NULL

iCount = 6

===============================================================================
Explanation:
===============================================================================

This program implements two important insertion operations of a Singly
Linked List.

InsertFirst():

• Creates a new node using the new operator.
• Initializes the data and next members.
• If the list is empty, the new node becomes the first node.
• Otherwise, the new node points to the current first node and then
  becomes the new first node.
• Increments iCount after successful insertion.

InsertLast():

• Creates a new node dynamically.
• If the list is empty, the new node becomes the first node.
• Otherwise, traverses the linked list until the last node is reached.
• Connects the last node to the new node.
• Increments iCount after insertion.

Display():

• Traverses the linked list from first.
• Prints every node until NULL is reached.

Count():

• Returns the value of iCount.
• Since iCount is updated after every insertion, counting is performed
  in constant time without traversing the linked list.

===============================================================================
Program Flow:
===============================================================================

Create SinglyLL Object
          │
          ▼
Constructor Invoked
          │
          ▼
Initialize first = NULL
Initialize iCount = 0
          │
          ▼
InsertFirst(51)
          │
          ▼
InsertFirst(21)
          │
          ▼
InsertFirst(11)
          │
          ▼
Display Linked List
          │
          ▼
Count Nodes
          │
          ▼
InsertLast(101)
          │
          ▼
InsertLast(111)
          │
          ▼
InsertLast(121)
          │
          ▼
Display Updated Linked List
          │
          ▼
Display Updated Count
          │
          ▼
Program Terminates

===============================================================================
Operations Performed:
===============================================================================

1. Created a Singly Linked List object.
2. Initialized first to NULL.
3. Initialized iCount to 0.
4. Inserted nodes at the beginning using InsertFirst().
5. Displayed the linked list.
6. Counted total nodes.
7. Inserted nodes at the end using InsertLast().
8. Displayed the updated linked list.
9. Displayed the updated node count.

===============================================================================
Time Complexity:
===============================================================================

InsertFirst()    : O(1)

InsertLast()     : O(n)

Display()        : O(n)

Count()          : O(1)

Space Complexity : O(1)
(Excluding dynamically allocated nodes)

===============================================================================
Conclusion:
===============================================================================

Program 011 demonstrates the implementation of both InsertFirst() and
InsertLast() operations in a Singly Linked List using C++. The program
illustrates dynamic node creation, pointer manipulation, list traversal,
and efficient node counting using the iCount data member. Together, these
operations form the foundation for more advanced linked list operations
such as insertion at a specific position and deletion.

===============================================================================
*/
